File: mobile-android/app/src/main/java/com/example/p2paudio/protocol/ConnectionCodeClient.kt
package com.example.p2paudio.protocol

import com.example.p2paudio.model.ConnectionCodePayload
import com.example.p2paudio.model.FailureCode
import com.example.p2paudio.model.SessionFailure
import java.io.IOException
import java.net.HttpURLConnection
import java.net.SocketTimeoutException
import java.net.URI
import java.net.URLEncoder
import java.nio.charset.StandardCharsets
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

class ConnectionCodeClientException(val failure: SessionFailure) : IOException(failure.message)

object ConnectionCodeClient {
    private const val CONNECT_TIMEOUT_MS = 10_000
    private const val READ_TIMEOUT_MS = 10_000

    suspend fun fetchInitPayload(connectionCode: ConnectionCodePayload): String = withContext(Dispatchers.IO) {
        val connection = openConnection(connectionCode, "/pairing/init").apply {
            requestMethod = "GET"
        }

        runRequest(
            connection = connection,
            successCode = HttpURLConnection.HTTP_OK,
            operationName = "fetch init payload"
        )
    }

    suspend fun submitConfirmPayload(connectionCode: ConnectionCodePayload, confirmPayload: String) = withContext(Dispatchers.IO) {
        val connection = openConnection(connectionCode, "/pairing/confirm").apply {
            requestMethod = "POST"
            doOutput = true
            setRequestProperty("Content-Type", "text/plain; charset=utf-8")
        }

        try {
            connection.outputStream.use { output ->
                output.write(confirmPayload.toByteArray(StandardCharsets.UTF_8))
            }
        } catch (error: IOException) {
            connection.disconnect()
            throw mapTransportFailure(error)
        }

        runRequest(
            connection = connection,
            successCode = HttpURLConnection.HTTP_ACCEPTED,
            operationName = "submit confirm payload"
        )
        Unit
    }

    private fun openConnection(connectionCode: ConnectionCodePayload, path: String): HttpURLConnection {
        val token = URLEncoder.encode(connectionCode.token, "UTF-8")
        val url = URI(
            "http",
            null,
            connectionCode.host,
            connectionCode.port,
            path,
            "token=$token",
            null
        ).toURL()

        return (url.openConnection() as HttpURLConnection).apply {
            connectTimeout = CONNECT_TIMEOUT_MS
            readTimeout = READ_TIMEOUT_MS
            useCaches = false
            setRequestProperty("Accept", "text/plain")
        }
    }

    private fun runRequest(
        connection: HttpURLConnection,
        successCode: Int,
        operationName: String
    ): String {
        try {
            val responseCode = connection.responseCode
            val body = readBody(connection, responseCode)
            if (responseCode == successCode) {
                return body
            }

            throw when (responseCode) {
                HttpURLConnection.HTTP_GONE -> ConnectionCodeClientException(
                    SessionFailure(
                        FailureCode.SESSION_EXPIRED,
                        "Connection code expired during $operationName"
                    )
                )
                HttpURLConnection.HTTP_BAD_REQUEST,
                HttpURLConnection.HTTP_UNAUTHORIZED,
                HttpURLConnection.HTTP_NOT_FOUND,
                HttpURLConnection.HTTP_CONFLICT -> ConnectionCodeClientException(
                    SessionFailure(
                        FailureCode.INVALID_PAYLOAD,
                        "Connection code was rejected during $operationName"
                    )
                )
                else -> ConnectionCodeClientException(
                    SessionFailure(
                        FailureCode.PEER_UNREACHABLE,
                        "Windows peer failed during $operationName: HTTP $responseCode ${body.ifBlank { "" }}".trim()
                    )
                )
            }
        } catch (failure: ConnectionCodeClientException) {
            throw failure
        } catch (error: IOException) {
            throw mapTransportFailure(error)
        } finally {
            connection.disconnect()
        }
    }

    private fun readBody(connection: HttpURLConnection, responseCode: Int): String {
        val inputStream = when {
            responseCode in 200..299 -> connection.inputStream
            connection.errorStream != null -> connection.errorStream
            else -> null
        } ?: return ""

        return inputStream.bufferedReader(StandardCharsets.UTF_8).use { reader ->
            reader.readText()
        }
    }

    private fun mapTransportFailure(error: IOException): ConnectionCodeClientException {
        if (error is SocketTimeoutException) {
            return ConnectionCodeClientException(
                SessionFailure(
                    FailureCode.PEER_UNREACHABLE,
                    "Timed out while contacting the Windows peer"
                )
            )
        }
        return ConnectionCodeClientException(
            SessionFailure(
                FailureCode.PEER_UNREACHABLE,
                error.message ?: "Failed to contact the Windows peer"
            )
        )
    }
}

File: desktop-windows/src/P2PAudio.Windows.App/Services/ConnectionCodeClient.cs
using System.Net;
using System.Net.Http;
using System.Text;
using P2PAudio.Windows.Core.Models;

namespace P2PAudio.Windows.App.Services;

public sealed class ConnectionCodeClientException : IOException
{
    public ConnectionCodeClientException(SessionFailure failure) : base(failure.Message)
    {
        Failure = failure;
    }

    public SessionFailure Failure { get; }
}

public static class ConnectionCodeClient
{
    private static readonly HttpClient HttpClient = new()
    {
        Timeout = TimeSpan.FromSeconds(10)
    };

    public static async Task<string> FetchInitPayloadAsync(
        ConnectionCodePayload connectionCode,
        CancellationToken cancellationToken = default)
    {
        using var request = CreateRequest(HttpMethod.Get, connectionCode, "/pairing/init");
        return await RunRequestAsync(request, HttpStatusCode.OK, "fetch init payload", cancellationToken);
    }

    public static async Task SubmitConfirmPayloadAsync(
        ConnectionCodePayload connectionCode,
        string confirmPayload,
        CancellationToken cancellationToken = default)
    {
        using var request = CreateRequest(HttpMethod.Post, connectionCode, "/pairing/confirm");
        request.Content = new StringContent(confirmPayload, Encoding.UTF8, "text/plain");
        _ = await RunRequestAsync(request, HttpStatusCode.Accepted, "submit confirm payload", cancellationToken);
    }

    private static HttpRequestMessage CreateRequest(
        HttpMethod method,
        ConnectionCodePayload connectionCode,
        string path)
    {
        var token = Uri.EscapeDataString(connectionCode.Token);
        var uri = new UriBuilder(Uri.UriSchemeHttp, connectionCode.Host, connectionCode.Port, path)
        {
            Query = $"token={token}"
        }.Uri;

        var request = new HttpRequestMessage(method, uri);
        request.Headers.Accept.ParseAdd("text/plain");
        request.Headers.ConnectionClose = true;
        return request;
    }

    private static async Task<string> RunRequestAsync(
        HttpRequestMessage request,
        HttpStatusCode successCode,
        string operationName,
        CancellationToken cancellationToken)
    {
        try
        {
            using var response = await HttpClient.SendAsync(
                request,
                HttpCompletionOption.ResponseHeadersRead,
                cancellationToken);

            return await ReadResponseBodyAsync(response, successCode, operationName, cancellationToken);
        }
        catch (ConnectionCodeClientException)
        {
            throw;
        }
        catch (HttpRequestException error)
        {
            throw MapTransportFailure(error);
        }
        catch (TaskCanceledException error) when (!cancellationToken.IsCancellationRequested)
        {
            throw MapTransportFailure(error);
        }
        catch (IOException error)
        {
            throw MapTransportFailure(error);
        }
    }

    private static async Task<string> ReadResponseBodyAsync(
        HttpResponseMessage response,
        HttpStatusCode successCode,
        string operationName,
        CancellationToken cancellationToken)
    {
        var body = response.Content is null
            ? string.Empty
            : await response.Content.ReadAsStringAsync(cancellationToken);

        if (response.StatusCode == successCode)
        {
            return body;
        }

        throw response.StatusCode switch
        {
            HttpStatusCode.Gone => new ConnectionCodeClientException(
                new SessionFailure(FailureCode.SessionExpired, $"Connection code expired during {operationName}")),
            HttpStatusCode.BadRequest or HttpStatusCode.Unauthorized or HttpStatusCode.NotFound or HttpStatusCode.Conflict
                => new ConnectionCodeClientException(
                    new SessionFailure(FailureCode.InvalidPayload, $"Connection code was rejected during {operationName}")),
            _ => new ConnectionCodeClientException(
                new SessionFailure(
                    FailureCode.PeerUnreachable,
                    $"Android peer failed during {operationName}: HTTP {(int)response.StatusCode} {body}".Trim()))
        };
    }

    private static ConnectionCodeClientException MapTransportFailure(Exception error)
    {
        if (error is TaskCanceledException or TimeoutException)
        {
            return new ConnectionCodeClientException(
                new SessionFailure(FailureCode.PeerUnreachable, "Timed out while contacting the Android peer"));
        }

        return new ConnectionCodeClientException(
            new SessionFailure(FailureCode.PeerUnreachable, error.Message));
    }
}
